File: models_py/b3_aging.py
"""B3 - aging curves (Python side of the parity gate).

Quadratic fit of OPS on age -> peak-age estimate. Cross-sectional over 2023-2025 (survivor bias
noted). Reads ONLY the gold feed.
"""
from __future__ import annotations

import json
import os

import pandas as pd
import statsmodels.api as sm


def main() -> None:
    gold = os.getenv("MLB_GOLD_DIR", "data/gold")
    res = os.getenv("MLB_RESULTS_DIR", "data/results")
    os.makedirs(res, exist_ok=True)

    df = pd.read_parquet(f"{gold}/player_age.parquet")
    df["age_sq"] = df["age"] ** 2
    m = sm.OLS(df["ops"].to_numpy(), sm.add_constant(df[["age", "age_sq"]].to_numpy())).fit()
    b0, b1, b2 = (float(c) for c in m.params)

    out = {
        "n": int(len(df)),
        "coef": {"intercept": round(b0, 8), "age": round(b1, 8), "age_sq": round(b2, 8)},
        "peak_age": round(-b1 / (2 * b2), 4),
        "r_squared": round(float(m.rsquared), 6),
    }
    with open(f"{res}/b3_aging__py.json", "w") as f:
        json.dump(out, f, indent=2)
    print(json.dumps({"stage": "models_py", "model": "b3_aging", **out}))


if __name__ == "__main__":
    main()

File: models_r/b3_aging.R
# B3 - aging curves (R side of the parity gate).
# Quadratic fit of OPS on age -> peak-age estimate (cross-sectional; survivor bias noted).
suppressMessages({
  library(arrow)
  library(dplyr)
  library(jsonlite)
})

gold <- Sys.getenv("MLB_GOLD_DIR", "data/gold")
res <- Sys.getenv("MLB_RESULTS_DIR", "data/results")
dir.create(res, showWarnings = FALSE, recursive = TRUE)

df <- read_parquet(file.path(gold, "player_age.parquet")) %>%
  mutate(age_sq = age^2)
m <- lm(ops ~ age + age_sq, data = df)
co <- coef(m)

out <- list(
  n = nrow(df),
  coef = list(
    intercept = round(co[["(Intercept)"]], 8),
    age = round(co[["age"]], 8),
    age_sq = round(co[["age_sq"]], 8)
  ),
  peak_age = round(-co[["age"]] / (2 * co[["age_sq"]]), 4),
  r_squared = round(summary(m)$r.squared, 6)
)

write_json(out, file.path(res, "b3_aging__r.json"), auto_unbox = TRUE, pretty = TRUE, digits = 12)
cat(toJSON(c(list(stage = "models_r", model = "b3_aging"), out), auto_unbox = TRUE, digits = 12), "\n")
